gtsfm/densify/patchmatchnet_data.py

```diff
@@ -0,0 +1,270 @@
+        pairs = np.argsort(-pair_scores, axis=1)[:, : self._num_views - 1]
+
+        # Filter out depth outliers and calculate depth ranges
+        depth_ranges = np.zeros((self._num_images_cal, 2))
+        for i in range(self._num_images_cal):
+            depth_ranges[i, 0] = np.percentile(depths[i], MIN_DEPTH_PERCENTILE)
+            depth_ranges[i, 1] = np.percentile(depths[i], MAX_DEPTH_PERCENTILE)
+
+        return pairs, depth_ranges
+
+    def __len__(self) -> int:
+        """Get the length of the dataset
+        In the PatchMatchNet workflow, each input image will be treated as a reference view once, and each input image
+        corresponds to a estimated camera pose. So the length of the dataset is equal to the number of reference views.
+
+        Returns:
+            the length of the dataset
+        """
+        return self._num_images_cal
+
+    def __getitem__(self, index: int) -> Dict[str, Any]:
+        """Get inference data to PatchmatchNet
+        Produces data containing _num_views images, the first images is the reference image
+
+        Args:
+            index: index of yield test data, the reference image ID of the test data is _keys[index]
+
+        Returns:
+            Dictionary containing:
+                "idx" test image index: int
+                "imgs" source and reference images: (num_views, image_channel, image_h, image_w)
+                "proj_matrices" projection matrices: (num_views, 4, 4)
+                "depth_min" minimum depth: int
+                "depth_max" maximum depth: int
+                "filename" output filename pattern: string
+        """
+        ref_key = self._patchmatchnet_idx_to_camera_idx[index]
+        src_keys = [self._patchmatchnet_idx_to_camera_idx[src_index] for src_index in self._pairs[index]]
+
+        cam_keys = [ref_key] + src_keys
+
+        imgs: List[np.ndarray] = [[] for _ in range(self._num_stages)]
+        proj_mats: List[np.ndarray] = [[] for _ in range(self._num_stages)]
+
+        for cam_key in cam_keys:
+            img = self._images[cam_key].value_array
+            np_img = np.array(img, dtype=np.float32) / 255.0
+            np_img = cv2.resize(np_img, (self._w, self._h), interpolation=cv2.INTER_LINEAR)
+            # Crop the image from the upper left corner, instead of from the center
+            #   to align with the 2D coordinates in track's measurements.
+            np_img = np_img[: self._cropped_h, : self._cropped_w, :]
+
+            intrinsics = self._sfm_result.get_camera(cam_key).calibration().K()
+            cTw = self._sfm_result.get_camera(cam_key).pose().inverse().matrix()
+            # In the multi-scale feature extraction, there are NUM_PATCHMATCHNET_STAGES stages.
+            #   The scales are [2^0, 2^(-1), 2^(-2), ..., 2^(1-NUM_PATCHMATCHNET_STAGES)]
+            #   Initially the intrinsics is scaled to fit the smallest image size
+            intrinsics[:2, :] /= 2 ** NUM_PATCHMATCHNET_STAGES
```

Comment:
can you add a comment summarizing whats happening here. Like we need images at different scales so doing xyz.